File: python/LinearNeedlemanWunsch.py
import numpy as np
from collections import deque
from SequenceAligner import SequenceAligner

# Performing Simple Needleman Wunsch
# Global Sequence Alignment

MATCH_MISMATCH_IDX = 0
QUERY_INSERTION_IDX = 1
QUERY_DELETION_IDX = 2

# Differences from Longest Common Subsequence --> Needleman-Wunsch
#   i. Now penalties are enforced (GAP and MISMATCH)
#  ii. Inclusion of different MATCH, MISMATCH, and GAP score

class LinearNeedlemanWunschAligner(SequenceAligner):
  
  def __init__(self, reference, query, MATCH, MISMATCH, GAP):
    super().__init__(reference, query)
    
    self.MATCH = MATCH         # Diagonal Matches
    self.MISMATCH = MISMATCH   # Diagonal Mismatches
    self.GAP = GAP             # Vertical/Horizontal Insertion/Deletions
    
  
  def execute(self):
    
    print("\nExecuting Linear Gap Penalty Needleman-Wunsch")
    print("Reference:", self.reference)
    print("Query:", self.query)
    print(f"(Match, Mismatch, Gap) = ({self.MATCH}, {self.MISMATCH}, {self.GAP})")
    
    # Initialize self.Memo correctly
    # self.Memo size (len(self.query) + 1) * (len(self.reference) + 1)
    self.initializeMemoMatrix()

    self.printMemoMatrix("Initialized Memo Matrix:")

    self.performRecursiveAnalysis()

    self.printMemoMatrix("Final Memo Matrix:")
    
    # Performing backtracking and printing out all the subsequences
    self.backtrackPrintAllPaths()


  # self.Memo size (len(self.query) + 1) * (len(self.reference) + 1)
  def initializeMemoMatrix(self) -> None:
    
    self.Memo = np.zeros((len(self.query) + 1, len(self.reference) + 1))
    
    for x_idx in range(0, len(self.reference)+1):
      self.Memo[0][x_idx] = x_idx * self.GAP
      
    for y_idx in range(0, len(self.query)+1):
      self.Memo[y_idx][0] = y_idx * self.GAP
    
    return
      

  # Fill the 2D self.Memo matrix
  def performRecursiveAnalysis(self) -> None:
    
    # Building a backtracking matrix to keep track of where it came from
    # Each element has 3 places it could backtrack to [corner, left, up]
    # Keeping track for each element whether or not it came from that location
    # It is possible to come from multiple locations (branches of backtracking)
    self.backtrackMatrix = np.zeros((len(self.query), len(self.reference), 3), dtype=bool)
    
    # Row 0 and column 0 already initialized
    
    # Going through all characters of the query sequence
    for query_idx in range(1, len(self.query) + 1):
      for reference_idx in range(1, len(self.reference) + 1):
        
        query_deletion_score = self.Memo[query_idx - 1][reference_idx] + self.GAP
        query_insertion_score = self.Memo[query_idx][reference_idx - 1] + self.GAP
        
        # If match, add the match score from the corner
        if (self.query[query_idx - 1] == self.reference[reference_idx - 1]):
          corner_score = self.Memo[query_idx - 1][reference_idx - 1] + self.MATCH
        # Otherwise, add the penalty score from the corner
        else:
          corner_score = self.Memo[query_idx - 1][reference_idx - 1] + self.MISMATCH
        
        self.Memo[query_idx][reference_idx] = max(query_deletion_score, query_insertion_score, corner_score)
        
        self.backtrackMatrix[query_idx-1][reference_idx-1][MATCH_MISMATCH_IDX] = (corner_score == self.Memo[query_idx][reference_idx])
        self.backtrackMatrix[query_idx-1][reference_idx-1][QUERY_INSERTION_IDX] = (query_insertion_score == self.Memo[query_idx][reference_idx])
        self.backtrackMatrix[query_idx-1][reference_idx-1][QUERY_DELETION_IDX] = (query_deletion_score == self.Memo[query_idx][reference_idx])  
      # end for col_idx
    # end for row_idx
    
    return
    

  # Performing backtracking to get longest subsequence
  def backtrackPrintAllPaths(self) -> None:
    
    # Appending the bottom right location to the queue
    # That will always be the starting point for needleman-wunsch
    # The index is pointing to the last element of the matrix, not of the strings
    currentReferenceIdx = len(self.reference)
    currentQueryIdx = len(self.query)
    trackerReference = ""
    trackerConnection = ""
    trackerQuery = ""
    
    pathsQueue = deque()
    
    pathsQueue.append([
      currentReferenceIdx, 
      currentQueryIdx, 
      trackerReference, 
      trackerConnection, 
      trackerQuery]
    )
    
    # While there are still paths to traverse
    while (pathsQueue):
      
      path = pathsQueue.popleft()
      
      currentReferenceIdx = path[0]
      currentQueryIdx = path[1]
      trackerReference = path[2]
      trackerConnection = path[3]
      trackerQuery = path[4]
      
      # If either of the indices are 0, we have traversed through the sequence
      if (not((currentReferenceIdx == 0) and (currentQueryIdx == 0))):
        
        if (self.backtrackMatrix[currentQueryIdx-1][currentReferenceIdx-1][MATCH_MISMATCH_IDX] and
            self.reference[currentReferenceIdx-1] == self.query[currentQueryIdx-1]):
          
          pathsQueue.append([
            currentReferenceIdx-1, 
            currentQueryIdx-1, 
            self.reference[currentReferenceIdx-1] + trackerReference, 
            "*" + trackerConnection, 
            self.query[currentQueryIdx-1] + trackerQuery]
          )
        #end if match
          
        if (self.backtrackMatrix[currentQueryIdx-1][currentReferenceIdx-1][MATCH_MISMATCH_IDX] and
            self.reference[currentReferenceIdx-1] != self.query[currentQueryIdx-1]):
          
          pathsQueue.append([
            currentReferenceIdx-1, 
            currentQueryIdx-1, 
            self.reference[currentReferenceIdx-1] + trackerReference, 
            "|" + trackerConnection, 
            self.query[currentQueryIdx-1] + trackerQuery]
          )
        #end if mismatch
          
        if (self.backtrackMatrix[currentQueryIdx-1][currentReferenceIdx-1][QUERY_INSERTION_IDX]):
          
          pathsQueue.append([
            currentReferenceIdx-1, 
            currentQueryIdx, 
            self.reference[currentReferenceIdx-1] + trackerReference, 
            " " + trackerConnection, 
            "_" + trackerQuery]
          )
        #end if left gap
          
        if (self.backtrackMatrix[currentQueryIdx-1][currentReferenceIdx-1][QUERY_DELETION_IDX]):
          
          pathsQueue.append([
            currentReferenceIdx, 
            currentQueryIdx-1, 
            "_" + trackerReference, 
            " " + trackerConnection, 
            self.query[currentQueryIdx-1] + trackerQuery]
          )
        #end if upper gap
      
      # end if trace not completed
          
      else:
        # Given Path Trace has reached  the border and is finished
        print()
        print(trackerReference)
        print(trackerConnection)
        print(trackerQuery)
        
    # end while
    print()
    
    return

File: c++/LinearNeedlemanWunsch.cpp
#include "LinearNeedlemanWunsch.h"
#include "FakeDPX.hpp"

using std::cout;
using std::vector;
using std::max;
using std::string;

void LinearNeedlemanWunsch::init_matrix(){
    cout << "[Initializing Matrix...]\n";
    size_t num_rows = query_str.size();
    size_t num_cols = reference_str.size();

    // Initialize scoring memo
    // Need the additional 0th row and column for initial gap penalty scores
    vector<int> temp_memo;
    temp_memo.resize(num_cols + 1, 0);
    memo.resize(num_rows + 1, temp_memo);

    // Initialize backtracking memo
    // As we are doing global alignment we need to be able to backtrack from 0th row and column
    vector<direction> temp_backtrack;
    temp_backtrack.resize(num_cols + 1, NONE);
    backtrack_memo.resize(num_rows + 1, temp_backtrack);

    // If query needed to do initial DELETIONS before entering reference
    // Can be thought of adding insertions to reference
    // As this is linear gap penalty, will be doing row_idx * gap_weight
    for (int row_idx = 1; row_idx < num_rows+1; ++row_idx) {
        memo[row_idx][0] = row_idx * gap_weight;
        backtrack_memo[row_idx][0] = QUERY_DELETION;
    }

    // If query needed to do initial INSERTIONS before entering reference
    // As this is linear gap penalty, will be doing row_idx * gap_weight
    for (int col_idx = 1; col_idx < num_cols+1; ++col_idx) {
        memo[0][col_idx] = col_idx * gap_weight;
        backtrack_memo[0][col_idx] = QUERY_INSERTION;
    }
}

void LinearNeedlemanWunsch::print_matrix(){
    
    size_t num_rows = query_str.size();
    size_t num_cols = reference_str.size();

    // Print parameters
    cout << "Reference: " << reference_str << " Size: " << num_cols << "\n";
    cout << "Query: " << query_str << " Size: " << num_rows << "\n";

    cout << "Matrix Dim: [ " << num_rows + 1 << " x " << num_cols + 1 << " ]\n";

    // Set the spacing for each element of the matrix
    int min_width = 2;

    // Print out reference str
    cout << "    " << std::setw(min_width) << " " << " ";
    for(size_t i = 0; i < reference_str.size(); i++){
        cout << "  " << std::setw(min_width) << reference_str[i] << " ";
    }
    cout << "\n";

    // Print rest of matrix
    for(size_t i = 0; i < memo.size(); i++){
        if(i == 0){
            cout << "  ";
        }
        else {
            cout << query_str[i-1] << " ";
        }
        cout << "[";
        for(size_t j = 0; j < memo[i].size(); j++){
            if(j != memo[i].size() - 1){
                cout << " " << std::setw(min_width) << memo[i][j] << ", ";
            }
            else {
                cout << " " << std::setw(min_width) << memo[i][j];
            }
        } //end for
        cout << "],\n";
    } //end for

    cout << std::endl;

}

void LinearNeedlemanWunsch::score_matrix() {
    cout << "[Scoring...]\n";

    bool pred;

    // Score the smith waterman matrix, keeping track of where a cell got its value
    for(size_t row_idx = 1; row_idx < memo.size(); row_idx++) {
        for(size_t col_idx = 1; col_idx < memo[0].size(); col_idx++) {
            
            int queryInsertionScore;
            int queryDeletionScore;
            int matchMismatchScore;
            int largestScore;
            
            direction corner_direction = NONE;

            // If match, add the match score from the corner
            if (query_str[row_idx - 1] == reference_str[col_idx - 1]){
                matchMismatchScore = memo[row_idx - 1][col_idx - 1] + match_weight;
                corner_direction = MATCH;
            }
            // Otherwise, add the penalty score from the corner
            else {
                matchMismatchScore = memo[row_idx - 1][col_idx - 1] + mismatch_weight;
                corner_direction = MISMATCH;
            }
            
            // Calculate potential gap scores
            queryDeletionScore = memo[row_idx - 1][col_idx] + gap_weight;
            queryInsertionScore = memo[row_idx][col_idx - 1] + gap_weight;
            
            largestScore = FakeDPX::__vibmax_s32(queryDeletionScore, matchMismatchScore, &pred);
            if (pred) corner_direction = QUERY_DELETION;
            
            largestScore = FakeDPX::__vibmax_s32(queryInsertionScore, largestScore, &pred);
            if (pred) corner_direction = QUERY_INSERTION;

            memo[row_idx][col_idx] = largestScore;
            backtrack_memo[row_idx][col_idx] = corner_direction;

        } // end for col_idx
    } // end for row_idx

    //cout << std::endl;
}

void LinearNeedlemanWunsch::backtrack(){
    cout << "[Finding max cells...]\n";

    // We need to find the max score of the entire matrix
    // + note the cell(s) that have this value

    int currentMemoRow = memo.size()-1;
    int currentMemoCol = memo[0].size()-1;

    std::string referenceSequence = "";
    std::string pairRelation = "";
    std::string querySequence = "";

    while ((currentMemoRow != 0) && (currentMemoCol != 0)) {
        
        // Determine the current cell's predecessor
        switch (backtrack_memo[currentMemoRow][currentMemoCol]) {
            
            case MATCH:
                referenceSequence = reference_str[currentMemoCol-1] + referenceSequence;
                pairRelation = "*" + pairRelation;
                querySequence = query_str[currentMemoRow-1] + querySequence;
                --currentMemoRow;
                --currentMemoCol;
                break;
            // end if match

            case MISMATCH: 
                referenceSequence = reference_str[currentMemoCol-1] + referenceSequence;
                pairRelation = "|" + pairRelation;
                querySequence = query_str[currentMemoRow-1] + querySequence;
                --currentMemoRow;
                --currentMemoCol;
                break;
            // end if mismatch
            
            case QUERY_DELETION:
                referenceSequence = "_" + referenceSequence;
                pairRelation = " " + pairRelation;
                querySequence = query_str[currentMemoRow-1] + querySequence;
                --currentMemoRow;
                break;
            // end if query deletion
            
            case QUERY_INSERTION:
                referenceSequence = reference_str[currentMemoCol-1] + referenceSequence;
                pairRelation = " " + pairRelation;
                querySequence = "_" + querySequence;
                --currentMemoCol;
                break;
            // end if query insertion
            
            default:
                exit(1);
            // end if upper gap

        } // end switch
    } // end while
    #ifdef PRINT_MATRIX
        print_matrix();
    #endif

    cout << "[Needleman-Wunsch Score: " << memo[memo.size()-1][memo[0].size()-1] << "]\n";
    cout << "[Sequence Pairing(s)]\n";
    cout << "====================\n";

    cout << referenceSequence << std::endl;
    cout << pairRelation << std::endl;
    cout << querySequence << std::endl;
    
    cout << "====================\n";
    cout << std::endl;
}

void LinearNeedlemanWunsch::align(){
    init_matrix();
    #ifdef PRINT_MATRIX
        print_matrix();
    #endif
    score_matrix();
    backtrack();
}

void LinearNeedlemanWunsch::print_results(){
//     cout << "[Scored Matrix]\n";
//     print_matrix();

//     cout << "[# Highest Scoring Sequences: " << results.size() << " Score: " << max_score << "]\n\n";
//     cout << "[Sequence Pairing(s)]\n";
//     cout << "\n====================\n";

//     for(size_t idx = 0; idx < results.size(); idx++){
//         cout << "Sequence Pair: " << idx << "\n";
//         backtrack_info backtrack_record = results[idx];
//         for(size_t base = 0; base < backtrack_record.query_sequence.size(); base++){
//             cout << backtrack_record.query_sequence[base];
//         }
//         cout << "\n";
//         for(size_t base = 0; base < backtrack_record.pair_relation.size(); base++){
//             cout << backtrack_record.pair_relation[base];
//         }
//         cout << "\n";
//         for(size_t base = 0; base < backtrack_record.reference_sequence.size(); base++){
//             cout << backtrack_record.reference_sequence[base];
//         }
//         cout << "\n====================\n";
//     }

//     cout << std::endl;
}


